runtime/interpreter/interpreter.go

```diff
@@ -1107,6 +1161,23 @@ func (interpreter *Interpreter) declareNonEnumCompositeValue(
 				locationRange := invocation.LocationRange
 				self := *invocation.Self
 
+				if len(compositeType.ConstructorParameters) < 1 {
+					return nil
+				}
+
+				// event interfaces do not exist
+				compositeDecl := declaration.(*ast.CompositeDeclaration)
+				if compositeDecl.IsResourceDestructionDefaultEvent() {
+					// we implicitly pass the containing composite value as an argument to this invocation
+					containerComposite := invocation.Arguments[0].(*CompositeValue)
+					interpreter.activations.PushNewWithParent(inter.activations.CurrentOrNew())
+					interpreter.SharedState.callStack.Push(invocation)
+					invocation.Arguments = interpreter.evaluateDefaultDestroyEvent(
```

Comment:
I had to re-read this piece a couple times to understand what is going on: There are two interpreters:
1. The one used when the event is declared, the "outer" `interpreter`
2. The one of the invocation, when the event is emitted, the "inner" `inter`

The declaration `interpreter` is used for evaluating the default arguments in the last line (https://github.com/onflow/cadence/pull/2817/files#diff-a36f730b18ca31316909bd74125003c1fbe785a13f926b4a0631e7c0cc6a51e4R1175). 

Is it correct to push the invocation `inter`'s current activation before though?

I'm worried that that is effectively dynamic instead of lexical scope. I think this works, because there is no way to actually refer to the `inter`'s activation's variables, because all the default argument expressions can only be constants or refer to `self`, and `base` in an attachment, right? The default arguments cannot refer to e.g. global variables?

Maybe add a test case that ensures the pushed activation is correct, e.g.

```cadence
let x = 1

resource R {
    event ResourceDestroyed(x: x)
}

fun test() {
    let x = 2
    let r <- create R()
    // should emit `R.ResourceDestroyed(x: 1)`, not `R.ResourceDestroyed(x: 2)`
    destroy r
}
```

Also, maybe rename `inter` to `invocationInter` or even shadow `interpreter`, to make it clearer what interpreter is what. It matters a lot here because of the scoping / activations.

Reply:
This test is not possible; `x` is not a valid argument to a default destroy event, since in general we cannot guarantee that an external variable access will not fail (e.g. an invalidated resource reference).

Reply:
I've added a variant of it with a field on `R` though.

Reply:
I renamed the variables to make it clearer what is going on though

Reply:
Could you please still add the example above as a test case? It would be good to ensure that the activation code is correct. 

In the test, allow the example above to type check, i.e. ignore the expected type checking error that is produced for `= x`. You can do this by using `parseCheckAndInterpretWithOptions` and 
```go
ParseCheckAndInterpretOptions{
	HandleCheckerError: func(err error) {
		errs := checker.RequireCheckerErrors(t, err, 1)
		// ...
	},
}
```

Reply:
ah okay, yeah I didn't know you could do that! Will add the test.

Reply:
Great test! Turns out neither `invocationInterpreter` nor `declarationInterpreter` were the right scoping for this; instead I captured the activations that exist at the time the event's host function is created, and manually put them on top of the stack for the duration of default argument interpreting. This makes the behavior correct I think. I've added a comment to this effect as well.